File: packages/core-py/src/ebb_ai/providers/gemini.py
"""Gemini adapter.

Talks to Google's Generative Language API
(https://generativelanguage.googleapis.com) directly over ``httpx`` — no
vendor SDK dependency. The API key is read from ``GEMINI_API_KEY``, falling
back to ``GOOGLE_API_KEY`` (the precedence Google's own ``google-genai`` SDK
uses).

Endpoint used:
  - ``POST /v1beta/models/{model}:generateContent`` — sync dispatch, returns
    text plus real ``usageMetadata`` token counts.

Batch: DELIBERATELY UNSUPPORTED (mirrors the TS adapter). The uniform batch
surface here is submit → batch_id → poll(retrieve_batch) → per-request
results, modelled on Anthropic Message Batches and OpenAI Batch Files. Gemini
does not map cleanly:

- Vertex AI batch prediction requires a GCS/BigQuery source/sink — a
  different I/O contract entirely.
- The Developer-API batch mode returns a long-running *operation* keyed by an
  operation name, not a batch id with a separate results endpoint, and its
  retrieval + usage semantics differ from the shared contract.

So this adapter implements only the sync path — it inherits the base
``dispatch_batch`` / ``retrieve_batch`` stubs (both raise). The scheduler
feature-detects that (:func:`ebb_ai.scheduler._has_batch_support`) and keeps
Gemini tasks on the sync path.
"""

from __future__ import annotations

import os
from typing import Any

import httpx

from .base import DispatchOptions, DispatchResult, ProviderAdapter

_DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com/v1beta"


class GeminiAdapter(ProviderAdapter):
    """Adapter for Google Gemini models via the Generative Language API.

    Parameters
    ----------
    api_key:
        Gemini API key. If omitted, read from ``GEMINI_API_KEY`` then
        ``GOOGLE_API_KEY``.
    base_url:
        Override the API base URL (defaults to the public endpoint).
    client:
        Optional pre-built ``httpx.AsyncClient`` (or any object exposing an
        awaitable ``post(url, *, headers, json)`` returning a response with
        ``status_code`` / ``json()`` / ``text``). Injected in tests to avoid
        live calls. When omitted, a short-lived client is created per call.
    timeout_s:
        Per-request timeout when constructing the default client.
    """

    name = "gemini"

    def __init__(
        self,
        *,
        api_key: str | None = None,
        base_url: str | None = None,
        client: Any | None = None,
        timeout_s: float = 60.0,
    ) -> None:
        self._api_key = (
            api_key
            or os.environ.get("GEMINI_API_KEY")
            or os.environ.get("GOOGLE_API_KEY")
        )
        self._base_url = (base_url or _DEFAULT_BASE_URL).rstrip("/")
        self._client = client
        self._timeout_s = timeout_s

    @property
    def ready(self) -> bool:
        """True once an API key is available (env or constructor)."""
        return bool(self._api_key)

    async def dispatch(
        self,
        model: str,
        prompt: str,
        options: DispatchOptions | None = None,
    ) -> DispatchResult:
        if not self._api_key:
            raise RuntimeError(
                "GeminiAdapter: no API key. Set GEMINI_API_KEY (or "
                "GOOGLE_API_KEY) or pass api_key=... to the constructor."
            )
        opts = options or DispatchOptions()

        generation_config: dict[str, Any] = {}
        if opts.max_tokens is not None:
            generation_config["maxOutputTokens"] = opts.max_tokens
        temperature = opts.extra.get("temperature")
        if temperature is not None:
            generation_config["temperature"] = temperature

        body: dict[str, Any] = {
            "contents": [{"role": "user", "parts": [{"text": prompt}]}],
        }
        if opts.system is not None:
            body["systemInstruction"] = {"parts": [{"text": opts.system}]}
        if generation_config:
            body["generationConfig"] = generation_config

        # The key rides the `x-goog-api-key` header (keeping it out of URLs /
        # access logs) rather than the legacy `?key=` query parameter.
        url = f"{self._base_url}/models/{model}:generateContent"
        headers = {
            "content-type": "application/json",
            "x-goog-api-key": self._api_key,
        }

        if self._client is not None:
            res = await self._client.post(url, headers=headers, json=body)
        else:
            async with httpx.AsyncClient(
                timeout=httpx.Timeout(self._timeout_s)
            ) as client:
                res = await client.post(url, headers=headers, json=body)

        if res.status_code != httpx.codes.OK:
            raise RuntimeError(f"Gemini API {res.status_code}: {res.text[:300]}")

        data = res.json()
        text = _extract_text(data)
        usage = data.get("usageMetadata") or {}
        return DispatchResult(
            text=text,
            model=data.get("modelVersion") or model,
            provider=self.name,
            raw=data,
            input_tokens=usage.get("promptTokenCount"),
            output_tokens=usage.get("candidatesTokenCount"),
        )


def _extract_text(data: Any) -> str:
    """Concatenate the text parts of the first candidate."""
    candidates = data.get("candidates") if isinstance(data, dict) else None
    if not candidates:
        return ""
    content = candidates[0].get("content") or {}
    parts = content.get("parts") or []
    chunks: list[str] = []
    for part in parts:
        value = part.get("text") if isinstance(part, dict) else None
        if isinstance(value, str):
            chunks.append(value)
    return "".join(chunks)


__all__ = ["GeminiAdapter"]

File: packages/core-ts/src/providers/gemini.ts
/**
 * Gemini provider adapter.
 *
 * Talks to Google's Generative Language API
 * (https://generativelanguage.googleapis.com) directly over `fetch` — no
 * vendor SDK dependency. The API key is read from `GEMINI_API_KEY`, falling
 * back to `GOOGLE_API_KEY` (the same precedence Google's own `@google/genai`
 * SDK uses).
 *
 * Endpoint used:
 *   - POST /v1beta/models/{model}:generateContent — sync dispatch, returns
 *     text plus real `usageMetadata` token counts.
 *
 * Batch: DELIBERATELY UNSUPPORTED. The uniform batch surface in this package
 * is submit → batchId → poll(retrieveBatch) → per-request results, modelled
 * on Anthropic Message Batches and OpenAI Batch Files (a flat 50% discount,
 * a discrete batch id, a distinct results endpoint, 24h SLA). Gemini's batch
 * options do not map cleanly:
 *   - Vertex AI batch prediction requires a GCS or BigQuery source/sink — a
 *     different I/O contract entirely (no inline prompt → text round-trip).
 *   - The Developer-API batch mode returns a long-running *operation* keyed
 *     by an operation name, not a batch id with a separate results endpoint,
 *     and its retrieval + usage semantics differ from the shared contract.
 * Rather than fake a mapping, this adapter omits `dispatchBatch` /
 * `retrieveBatch`. The scheduler feature-detects that and keeps Gemini tasks
 * on the sync path (batch-incapable adapters are a first-class case).
 */

import type {
  DispatchOptions,
  DispatchResult,
  ProviderAdapter,
} from "./base.js";

/** Minimal shape of the `fetch` we depend on (injectable for tests). */
type FetchLike = (
  url: string,
  init: {
    method: string;
    headers: Record<string, string>;
    body: string;
  },
) => Promise<{
  ok: boolean;
  status: number;
  text(): Promise<string>;
  json(): Promise<unknown>;
}>;

export interface GeminiAdapterOptions {
  /**
   * API key. Defaults to `process.env.GEMINI_API_KEY`, then
   * `process.env.GOOGLE_API_KEY`.
   */
  apiKey?: string;
  /** Override the API base URL (defaults to the public endpoint). */
  baseUrl?: string;
  /** Inject a `fetch` implementation. Defaults to the global `fetch`. */
  fetchImpl?: FetchLike;
}

const DEFAULT_BASE_URL = "https://generativelanguage.googleapis.com/v1beta";

export class GeminiAdapter implements ProviderAdapter {
  readonly provider = "gemini";
  private readonly apiKey: string | undefined;
  private readonly baseUrl: string;
  private readonly fetchImpl: FetchLike | undefined;

  constructor(opts: GeminiAdapterOptions = {}) {
    this.apiKey =
      opts.apiKey ??
      process.env.GEMINI_API_KEY ??
      process.env.GOOGLE_API_KEY;
    this.baseUrl = (opts.baseUrl ?? DEFAULT_BASE_URL).replace(/\/$/, "");
    this.fetchImpl = opts.fetchImpl;
  }

  get ready(): boolean {
    return Boolean(this.apiKey);
  }

  async dispatch(
    model: string,
    prompt: string,
    options: DispatchOptions = {},
  ): Promise<DispatchResult> {
    if (!this.apiKey) {
      throw new Error(
        "GeminiAdapter: no API key. Set GEMINI_API_KEY (or GOOGLE_API_KEY) or pass { apiKey } to the constructor.",
      );
    }
    const fetchImpl = this.fetchImpl ?? (globalThis.fetch as FetchLike | undefined);
    if (typeof fetchImpl !== "function") {
      throw new Error(
        "GeminiAdapter: no fetch implementation available. Pass { fetchImpl } or run on a runtime with global fetch.",
      );
    }

    const generationConfig: Record<string, unknown> = {};
    if (options.maxTokens !== undefined) {
      generationConfig.maxOutputTokens = options.maxTokens;
    }
    if (options.temperature !== undefined) {
      generationConfig.temperature = options.temperature;
    }

    const body: Record<string, unknown> = {
      contents: [{ role: "user", parts: [{ text: prompt }] }],
    };
    if (options.system) {
      body.systemInstruction = { parts: [{ text: options.system }] };
    }
    if (Object.keys(generationConfig).length > 0) {
      body.generationConfig = generationConfig;
    }

    // The API key rides in the `x-goog-api-key` header (keeping it out of the
    // URL / access logs) rather than the legacy `?key=` query parameter.
    const url = `${this.baseUrl}/models/${encodeURIComponent(model)}:generateContent`;
    const res = await fetchImpl(url, {
      method: "POST",
      headers: {
        "content-type": "application/json",
        "x-goog-api-key": this.apiKey,
      },
      body: JSON.stringify(body),
    });
    if (!res.ok) {
      let detail = "";
      try {
        detail = (await res.text()).slice(0, 300);
      } catch {
        detail = "(no response body)";
      }
      throw new Error(`Gemini API ${res.status}: ${detail}`);
    }
    const data = (await res.json()) as {
      candidates?: Array<{
        content?: { parts?: Array<{ text?: string }> };
      }>;
      usageMetadata?: {
        promptTokenCount?: number;
        candidatesTokenCount?: number;
        totalTokenCount?: number;
      };
      modelVersion?: string;
    };
    const text = (data.candidates?.[0]?.content?.parts ?? [])
      .map((p) => (typeof p.text === "string" ? p.text : ""))
      .join("");
    const inputTokens = data.usageMetadata?.promptTokenCount;
    const outputTokens = data.usageMetadata?.candidatesTokenCount;
    const totalTokens =
      data.usageMetadata?.totalTokenCount ??
      (inputTokens !== undefined && outputTokens !== undefined
        ? inputTokens + outputTokens
        : undefined);
    return {
      text,
      usage: { inputTokens, outputTokens, totalTokens },
      model: data.modelVersion ?? model,
      provider: this.provider,
      raw: data,
    };
  }

  // No dispatchBatch / retrieveBatch — see the module docstring. The
  // scheduler keeps Gemini tasks on the sync path.
}
